csrc/lower_thread_predicate.cpp

```diff
@@ -494,6 +519,224 @@ class RedundantUseAnalysis : BackwardVisitor {
+      }
+    }
+    // The following sort is added because in NVFuserTest.FusionIssue2076_CUDA
+    // the order is [B2, I1, I3] while the correct order should be [I3, B2, I1]
+    int n_elements = merged_root_domains.size();
+    std::vector<int> indices(n_elements);
+    std::iota(indices.begin(), indices.end(), 0);
+    std::sort(indices.begin(), indices.end(), [&](int a, int b) {
+      return index_root_domain[a] > index_root_domain[b];
+    });
+    std::vector<IterDomain*> merged_root_domains_sorted(n_elements);
+    for (int i = 0; i < n_elements; ++i) {
+      merged_root_domains_sorted[i] = merged_root_domains[indices[i]];
+    }
+
+    return merged_root_domains_sorted;
+  };
+
+  // Set stride when use extent of merged root domains to index leaf domain.
+  // e.g. Root: [I1,B2,B3] -> Leaf: [I1B2B3], the merged_root_domains =
+  // {B3,B2,I1}. root_stride = {1, len(B3), len(B3) * len(B2)}. For broadcast
+  // root domain, use the extent of its concretized domain.
+  auto getRootStride = [&](const std::vector<IterDomain*>& merged_root_domains,
+                           const std::unordered_map<IterDomain*, IterDomain*>&
+                               concretized_broadcast_root_domains) {
+    std::vector<Val*> root_stride(merged_root_domains.size());
+    root_stride[0] = IrBuilder::create<Int>(1);
+    for (int i = 1; i < (int)merged_root_domains.size(); i++) {
+      auto pre_crd = merged_root_domains[i - 1];
+      Val* pre_extent = pre_crd->isBroadcast()
+          ? concretized_broadcast_root_domains.at(pre_crd)->extent()
+          : pre_crd->extent();
+      root_stride[i] = IrBuilder::mulExpr(root_stride[i - 1], pre_extent);
+    }
+    return root_stride;
+  };
+
+  // set write stride
+  // e.g. Root: [I1,B2,B3] -> Leaf: [I1B2B3], the merged_root_domains =
+  // {B3,B2,I1}. root_stride = {1, len(B3), len(B3) * len(B2)}. write_stride_mod
+  // = {len(B3) * len(B2)}. More cases can be found in
+  // NVFuserTest.FusionAvoidRedundantWrite_CUDA
+  auto setWriteStride = [&](const std::vector<IterDomain*>& merged_root_domains,
+                            const std::vector<Val*>& root_stride,
+                            const ParallelType& pt) {
+    int ndim = (int)merged_root_domains.size();
+    if (ndim > 3) {
+      // only tested up to 3 dims merged into 1 dim
+      return;
+    }
+
+    // backward iterator, to ensure the visit is from outer to inner dims
+    int idx = ndim - 1;
+    while (idx >= 0) {
+      auto crd = merged_root_domains.at(idx);
+      if (crd->isBroadcast()) {
+        // for pattern [B1,X], only first len(X) blocks needs to write, len(X) =
+        // root_stride[idx of crd]
+        thread_predicates_[out_tv].write_stride_less[pt] = root_stride[idx];
+        thread_predicates_[out_tv].redundant_types.set(pt);
+        idx--;
+      } else if (crd->isIteration()) {
+        // for pattern [I1,B2,I3], write len(I3) blocks for every len[B2] *
+        // len[I3] blocks
```

Comment:
This is confusing to me, can you expand the comments in this section a bit?

Reply:
I changed the algorithm.
```
// e.g. [I2,I4] -> [B1,I2,B3,I4,B5] -> [B1*I2*B3*I4*B5]_blockIdx.x where Bx are concretized with extents larger than 1.
// In this case, the leaf domain {B1*I2*B3*I4*B5} is written redundantly because len(B1)*len(I2)*len(B3)*len(I4)*len(B5) blocks are writing to len(I2)*len(I4) locations.
// This class will set a map from ParallelType to the index that needs to write.
// The method is: (1) Thinking the leaf domain as a 5-D array with extended_stride:
// extended_stride = [len(I2)*len(B3)*len(I4)*len(B5), len(B3)*len(I4)*len(B5), len(I4)*len(B5), len(B5), 1]
// (2) calculate the index of each dimension: linear_index = blockIdx.x; index[i] = (linear_index / extended_stride[i]); linear_index %= extended_stride[i];
// (3) only the non-broadcasted dimensions need to write, dim-1(I2) and dim-3(I4). So the write index is:
// write_index = index[1] * extended_stride[1] + index[3] * extended_stride[3] 
// Finally, the write condition is: if (blockIdx.x == write_index)
```

Reply:
I forgot to push my local change on last Monday, updated now.